org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLFormatterTest.java

```diff
@@ -2055,6 +2054,150 @@ public void testNoSpacesOnNewLine() throws BadLocationException {
 		format(content, expected, formattingOptions);
 	}
 
+	@Test
+	public void testDontInsertFinalNewLine1() throws BadLocationException {
+		XMLFormattingOptions formattingOptions = createDefaultFormattingOptions();
+		formattingOptions.setTrimFinalNewlines(false);
+		formattingOptions.setInsertFinalNewline(true);
+		String content = "";
+		format(content, content, formattingOptions);
+	}
+
+	@Test
+	public void testDontInsertFinalNewLine2() throws BadLocationException {
+		XMLFormattingOptions formattingOptions = createDefaultFormattingOptions();
+		formattingOptions.setTrimFinalNewlines(false);
+		formattingOptions.setInsertFinalNewline(true);
+		String content = "<a  ></a>" + lineSeparator();
+		String expected = "<a></a>" + lineSeparator();
+		format(content, expected, formattingOptions);
+	}
+
+	@Test
+	public void testDontInsertFinalNewLine3() throws BadLocationException {
+		XMLFormattingOptions formattingOptions = createDefaultFormattingOptions();
+		formattingOptions.setTrimFinalNewlines(false);
+		formattingOptions.setInsertFinalNewline(true);
+		String content = "<a  ></a>" + lineSeparator() + "   ";
+		String expected = "<a></a>" + lineSeparator() + "   ";;
+		format(content, expected, formattingOptions);
+	}
+
+	@Test
+	public void testInsertFinalNewLine1() throws BadLocationException {
+		XMLFormattingOptions formattingOptions = createDefaultFormattingOptions();
+		formattingOptions.setInsertFinalNewline(true);
+		String content = "<a></a>";
+		String expected = "<a></a>" + lineSeparator();
+		format(content, expected, formattingOptions);
+	}
+
+	@Test
+	public void testInsertFinalNewLine2() throws BadLocationException {

```

Comment:
I will add the same test but by forcing line separator in the XML.

```
String content = "<a></a>\r\n\r\n";
String expected = "<a></a>\r\n";
```

to check that insert final new line use the lien separator declared in the XML (and not the line separator from the OS)